### DISCUSSION TITLE
Orin NX GStreamer Hardware encoder on scarthgap

### DISCUSSION THREAD

**Discussion by graugans** [OE4T/meta-tegra#1791] [Q&A] [answered]
Hej,

I do have a question. When building for the Orin NX for scarthgap I do not manage to get a hardware encoder for the h264 video. What am I doing wrong? The following packages are installed:

```
    gstreamer1.0 \
    gstreamer1.0-plugins-bad \
    gstreamer1.0-plugins-bad \
    gstreamer1.0-plugins-bad-videoparsersbad \
    gstreamer1.0-plugins-base \
    gstreamer1.0-plugins-base-rawparse \
    gstreamer1.0-plugins-base-tcp \
    gstreamer1.0-plugins-base-videorate \
    gstreamer1.0-plugins-base-videotestsrc \
    gstreamer1.0-plugins-good \
    gstreamer1.0-plugins-good \
    gstreamer1.0-plugins-good-imagefreeze \
    gstreamer1.0-plugins-good-isomp4 \
    gstreamer1.0-plugins-good-jpeg \
    gstreamer1.0-plugins-good-png \
    gstreamer1.0-plugins-good-rtp \
    gstreamer1.0-plugins-good-rtsp \
    gstreamer1.0-plugins-good-udp \
    gstreamer1.0-plugins-good-video4linux2 \
    gstreamer1.0-plugins-good-videocrop \
    gstreamer1.0-plugins-nvarguscamerasrc \
    gstreamer1.0-plugins-nvcompositor \
    gstreamer1.0-plugins-nvdrmvideosink \
    gstreamer1.0-plugins-nveglgles \
    gstreamer1.0-plugins-nvjpeg \
    gstreamer1.0-plugins-nvtee \
    gstreamer1.0-plugins-nvv4l2camerasrc \
    gstreamer1.0-plugins-nvvidconv \
    gstreamer1.0-plugins-nvvideo4linux2 \
    gstreamer1.0-plugins-nvvideosinks \
    gstreamer1.0-plugins-tegra \
    gstreamer1.0-rtsp-server \
    nvgstapps \
```

But when I try to query the gst features I do get:
```
# gst-inspect-1.0   | grep ^nv 
nvarguscamerasrc:  nvarguscamerasrc: NvArgusCameraSrc
nvcompositor:  nvcompositor: NvCompositor
nvdrmvideosink:  nvdrmvideosink: Nvidia Drm Video Sink
nveglglessink:  nveglglessink: EGL/GLES vout Sink
nveglstreamsrc:  nveglstreamsrc: nVidia EGL Stream
nvegltransform:  nvegltransform: NvEGLTransform
nvipcpipeline:  nvipcpipelinesink: Inter-process Pipeline Sink
nvipcpipeline:  nvipcpipelinesrc: Inter-process Pipeline Source
nvipcpipeline:  nvipcslavepipeline: Inter-process slave pipeline
nvivafilter:  nvivafilter: NvIVAFilter Plugin
nvjpeg:  nvjpegdec: JPEG image decoder
nvjpeg:  nvjpegenc: JPEG image encoder
nvtee:  nvtee: NvTee
nvunixfd:  nvunixfdsink: Unix file descriptor sink
nvunixfd:  nvunixfdsrc: Unix file descriptor source
nvv4l2camerasrc:  nvv4l2camerasrc: NvV4l2CameraSrc
nvvidconv:  nvvidconv: NvVidConv Plugin
nvvideo4linux2:  nvv4l2decoder: NVIDIA v4l2 video decoder
nvvideosink:  nvvideosink: nVidia Video Sink
nvvideosinks:  nv3dsink: Nvidia 3D sink
```

I would expect something like mentioned [here](https://forums.developer.nvidia.com/t/orin-nx-hardware-encoder-gstreamer-elements/288644/7):

``
nvvideo4linux2:  nvv4l2av1enc: V4L2 AV1 Encoder
nvvideo4linux2:  nvv4l2vp9enc: V4L2 VP9 Encoder
nvvideo4linux2:  nvv4l2h265enc: V4L2 H.265 Encoder
nvvideo4linux2:  nvv4l2h264enc: V4L2 H.264 Encoder
nvvideo4linux2:  nvv4l2decoder: NVIDIA v4l2 video decoder
```

Any ideas?

**Comment by dwalkes** [comment]
Short answer is I don't know what the issues is and I haven't seen this before myself.

I think including `gstreamer1.0-plugins-nvvideo4linux2` should be all you need to do from a build perspective as it should be built and installed as a part of this recipe.

I don't understand what the forum thread is recommending regarding the i2cdump checks but perhaps this is related, not sure if you've tried the suggestions there.

I think I'd try reproducing or getting a working configuration with the same hardware config and the stock release from NVIDIA and compare with the meta-tegra build, assuming you can't reproduce with the same release from NVIDIA.

**Reply by graugans** [reply]
Hej,

thanks for the response.Yes the forum entry is not very helpful, this was the only meaningful I could find. When calling the `i2cdetect` command I got Devive or resource bussy or something similar. I assume this is mostly to check weather a driver is loaded or not.

I am a bit overwhelmed by the nested pacakges and the different unwrapping of the binary packages. I guess this package is providing the plugins. [tegra-libraries-multimedia-v4l](https://github.com/OE4T/meta-tegra/blob/scarthgap/recipes-bsp/tegra-binaries/tegra-libraries-multimedia-v4l_36.4.0.bb#L19) within I only see the deployment of the decoder:

```sh
TEGRA_PLUGINS = "\
    libv4l2_nvvideocodec.so \
    libv4l2_nvcuvidvideocodec.so \
"
```

The only reference for the encoder I could find is in [tegra-libraries-multimedia](https://github.com/OE4T/meta-tegra/blob/scarthgap/recipes-bsp/tegra-binaries/tegra-libraries-multimedia_36.4.0.bb)

```sh
TEGRA_LIBRARIES_TO_INSTALL = "\
...
    nvidia/libnvvideoencode_ppe.so \
...
"
```
I also have checked the community layer but this only contains decoding as-well [gstreamer-tests)](https://github.com/OE4T/meta-tegra-community/tree/scarthgap/recipes-test/tegra-tests/gstreamer-tests)

At the moment I have converted my Orin Nano DevKit to a Orin NX DevKit by removing the Nano and replacing it with an Orin NX. Maybe I have missed some DeviceTree settings or Kernel config. Any Hints would be highly appreciated.

**Reply by madisongh** [reply]
Did you re-flash with a `p3768-0000-p3767-0000` image (or ...`-0001` for NX 8GB) when you swapped in the new module? Since the Nano does not support the hardware video encoder, it's possible that a kernel/device tree built for the Nano might have the encoder device disabled, and that could cause it not to show up in the gstreamer plugin.

**Reply by manvarpratik** [reply, marked as answer]
The device tree compatible property was not correct.

The gstreamer library (libgstnvvideo4linux2.so) identifies the device by the device tree `compatible` property `/sys/firmware/devicetree/base/compatible`.

As per the [Nvidia Documentation](https://docs.nvidia.com/jetson/archives/r35.3.1/DeveloperGuide/index.html#devices-supported-by-this-document), these are device codes,
.P3767-003,P3767-004,P3767-005  -- Orin Nano
.P3767-000,P3767-001 -- Orin NX
.P3701-000,P3701-004,P3701-005 -- Orin AGX

So, need to correct the base device tree compatible string in device tree file.
